zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/state/DbKeyGenerator.java

```diff
@@ -31,14 +32,20 @@ public final class DbKeyGenerator implements KeyGeneratorControls {
    */
   public DbKeyGenerator(
       final int partitionId, final ZeebeDb zeebeDb, final TransactionContext transactionContext) {
+    this.partitionId = partitionId;
     keyStartValue = Protocol.encodePartitionId(partitionId, INITIAL_VALUE);
     nextValueManager =
         new NextValueManager(keyStartValue, zeebeDb, transactionContext, ZbColumnFamilies.KEY);
   }
 
   @Override
   public long nextKey() {
-    return nextValueManager.getNextValue(LATEST_KEY);
+    final var nextKey = nextValueManager.getNextValue(LATEST_KEY);
+    if (Protocol.decodePartitionId(nextKey) != partitionId) {
+      throw new IllegalStateException(
+          String.format("Generated key %d does not belong to partition %d", nextKey, partitionId));
+    }
+    return nextKey;
```

Comment:
This might be not enough, as this will just fail the processing of the command and banning the corresponding instance. What we want is to hard fail right?

Reply:
Yeah I think so 👍🏼